ehrql/quiz.py

```diff
@@ -0,0 +1,264 @@
+from typing import Any
+
+from ehrql.query_engines.in_memory_database import (
+    EventColumn,
+    EventTable,
+    PatientColumn,
+    PatientTable,
+)
+from ehrql.query_engines.sandbox import EmptyDataset, SandboxQueryEngine
+from ehrql.query_language import BaseFrame, BaseSeries, Dataset
+
+
+def check_answer(
+    engine: SandboxQueryEngine, answer: Any, expected: Dataset | BaseFrame | BaseSeries
+) -> str:
+    message = check_type(answer, expected)
+    if message:
+        return message
+
+    ev_answer = evaluate(engine, answer)
+    ev_expected = evaluate(engine, expected)
+
+    message = check_type(ev_answer, ev_expected)
+    if message:
+        return message
+
+    if ev_answer == ev_expected:
+        return "Correct!"
+
+    checks = [
+        check_dataset_not_empty,
+        check_dataset_columns,
+        check_patient_ids,
+        check_patient_table_values,
+        check_patient_column_values,
+        check_event_row_ids,
+        check_event_table_values,
+        check_event_column_values,
+    ]
+
+    for check in checks:
+        message = check(ev_answer, ev_expected)
+        if message is None:
+            continue  # to the next check
+        return message
+    return "\n".join(
+        [
+            "Incorrect answer.",
+            "Expected:",
+            str(ev_expected),
+            "Got:",
+            str(ev_answer),
+        ]
+    )
+
+
+def get_type_name(obj: Any) -> str:
+    # Simplify the type name for better error messages
+    if isinstance(obj, BaseFrame):
+        return "Table"
+    elif isinstance(obj, BaseSeries):
+        return "Series"
+    return type(obj).__name__
+
+
+def check_type(answer: Any, expected: Any) -> str:
+    if not isinstance(answer, type(expected)):
+        answer_type_name = get_type_name(answer)
+        expected_type_name = get_type_name(expected)
+
+        if answer_type_name == "EmptyDataset":
+            return None  # Return a different message for empty datasets
+        if answer_type_name != expected_type_name:
+            # Only return an error message if it is helpful
+            return f"Expected {expected_type_name}, got {answer_type_name} instead."
+    return None
+
+
+def evaluate(
+    engine: SandboxQueryEngine, answer: Dataset | BaseFrame | BaseSeries
+) -> Any:
+    if isinstance(answer, Dataset):
+        return engine.evaluate_dataset(answer)
+    return engine.evaluate(answer)
+
+
+def check_dataset_not_empty(ev_ans: Any, ev_exp: Any) -> str | None:
```

Comment:
Nitpick: it wasn't immediately obvious to me what `ev_ans` and `ev_exp` meant. How annoying would it be to expand them out to `ev_answer` and `ev_expected` / `ev_expectation` (here and elsewhere)?

Reply:
Done! Thanks :)